File: src/main/java/org/library/Book.java
package org.library;

import java.util.Queue;
import java.util.LinkedList;

public class Book {
    private BookDetails bookDetails;
    private LibraryDate dueDate;
    private Queue<User> holdQueue;
    private User curHolder;
    private User curBorrower;

    Book(String title, String author){
        bookDetails = new BookDetails(title, author);
        dueDate = null;
        holdQueue = new LinkedList<>();
        curHolder = null;
        curBorrower = null;
    }

    public void addHolder(User user){
        holdQueue.add(user);
    }

    public User popHolder(){
        return holdQueue.poll();
    }

    public User removeCurHolder(){
        User holder = curHolder;
        curHolder = null;
        return holder;
    }

    public User removeCurBorrower(){
        User borrower = curBorrower;
        curBorrower = null;
        return borrower;
    }

    public AvailabilityEnum getAvailabilityStatus(String username){
        if (hasHolder() && !hasBorrower() && curHolder.getUsername().equals(username))
            return AvailabilityEnum.AVAILABLE;
        if (hasBorrower())
            return AvailabilityEnum.CHECKED_OUT;
        if (hasHolder())
            return AvailabilityEnum.ON_HOLD;
        return AvailabilityEnum.AVAILABLE;
    }

    // getters
    public String getTitle(){ return bookDetails.getTitle(); }
    public String getAuthor() { return bookDetails.getAuthor(); }
    public LibraryDate getDueDate() {
        if (dueDate == null) return null;
        return new LibraryDate(dueDate);
    }

    public int getHoldersNum(){ return holdQueue.size(); }
    public User peekHolderQueue(){
        return holdQueue.peek();
    }
    public User getCurHolder(){ return curHolder; }
    public BookDetails getBookDetails(){ return bookDetails; }
    public User getCurBorrower(){ return curBorrower; }
    public String getDueDateStr(){
        if (dueDate == null) return "N/A";
        return dueDate.toString();
    }

    // setters
    public void setDueDate(LibraryDate date){ dueDate = date; }
    public void setCurHolder(User holder){ curHolder = holder; }
    public void setCurBorrower(User borrower){ curBorrower = borrower; }

    // booleans
    public boolean containsHolder(User user){ return holdQueue.contains(user); }
    public boolean hasBorrower(){ return curBorrower != null; }
    public boolean hasHolder(){ return curHolder != null; }
    public boolean curBorrowerIs(String username){
        if (!hasBorrower()) return false;
        return username.equals(curBorrower.getUsername());
    }
    public boolean curHolderIs(String username){
        if (!hasHolder()) return false;
        return username.equals(curHolder.getUsername());
    }

    @Override
    public String toString() {
        return String.format("%s by %s", getTitle(), getAuthor());
    }
}

File: main/Book.js
import BookDetails from "./BookDetails.js";
import { AvailabilityEnum } from "./AvailabilityEnum.js";

export default class Book {
  constructor(title, author) {
    this.bookDetails = new BookDetails(title, author);
    this.dueDate = null;
    this.holdQueue = [];
    this.curBorrower = null;
  }

  addHolder(user) {
    this.holdQueue.push(user);
  }

  popHolder() {
    return this.holdQueue.shift() || null;
  }

  removeCurBorrower() {
    this.curBorrower = null;
    return null;
  }

  getAvailabilityStatus(username) {
    if (this.hasHolder() && !this.hasBorrower() && this.curHolderIs(username))
      return AvailabilityEnum.AVAILABLE;
    if (this.hasBorrower()) return AvailabilityEnum.CHECKED_OUT;
    if (this.hasHolder()) return AvailabilityEnum.ON_HOLD;
    return AvailabilityEnum.AVAILABLE;
  }

  // getters
  getTitle() {
    return this.bookDetails.getTitle();
  }
  getAuthor() {
    return this.bookDetails.getAuthor();
  }
  getDueDate() {
    return this.dueDate === null ? null : this.dueDate;
  }
  getHoldersNum() {
    return this.holdQueue.length;
  }
  getCurHolder() {
    return this.holdQueue.length ? this.holdQueue[0] : null;
  }
  getBookDetails() {
    return this.bookDetails;
  }
  getCurBorrower() {
    return this.curBorrower;
  }
  getDueDateStr() {
    return this.dueDate == null ? "N/A" : String(this.dueDate);
  }
  getQueue() {
    return this.holdQueue;
  }

  // setters
  setDueDate(date) {
    this.dueDate = date;
  }
  setCurBorrower(borrower) {
    this.curBorrower = borrower;
  }

  // booleans
  containsHolder(user) {
    return this.holdQueue.some((u) =>
      typeof u.equals === "function" ? u.equals(user) : u === user
    );
  }
  hasBorrower() {
    return this.curBorrower != null;
  }
  hasHolder() {
    return this.holdQueue.length > 0;
  }
  curBorrowerIs(username) {
    if (!this.hasBorrower()) return false;
    return username === this.curBorrower.getUsername();
  }
  curHolderIs(username) {
    if (!this.hasHolder()) return false;
    const h = this.getCurHolder();
    return (
      h && typeof h.getUsername === "function" && username === h.getUsername()
    );
  }

  toString() {
    return `${this.getTitle()} by ${this.getAuthor()}`;
  }
}
